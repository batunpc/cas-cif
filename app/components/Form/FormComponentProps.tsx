import type { AjvError } from "@rjsf/core";
export default interface FormComponentProps {
  formData: any;
  onChange: (changeObject: object) => void;
  onFormErrors: (errorsArray: AjvError[]) => void;
}
