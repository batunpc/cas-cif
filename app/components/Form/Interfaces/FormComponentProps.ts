import { FormProps } from "@rjsf/core";

type FormComponentProps<T = any> = Partial<FormProps<T>>;

export default FormComponentProps;
