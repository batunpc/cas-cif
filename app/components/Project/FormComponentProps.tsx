export default interface FormComponentProps {
  formData: any;
  applyChangesFromComponent: (changeObject: object) => void;
  onFormErrors: (errorsObject: object) => void;
}
