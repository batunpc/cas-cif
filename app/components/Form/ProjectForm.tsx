import type { JSONSchema7 } from "json-schema";
import FormBase from "../Form/FormBase";
import FormComponentProps from "../Form/FormComponentProps";
import { graphql, useFragment } from "react-relay";
import type { ProjectForm_query$key } from "__generated__/ProjectForm_query.graphql";
import { useMemo } from "react";
import SelectRfpWidget from "components/Form/SelectRfpWidget";
import SelectProjectStatusWidget from "./SelectProjectStatusWidget";

interface Props extends FormComponentProps {
  query: ProjectForm_query$key;
}

const ProjectForm: React.FC<Props> = (props) => {
  const { query } = useFragment(
    graphql`
      fragment ProjectForm_query on Query {
        query {
          allOperators {
            edges {
              node {
                rowId
                legalName
                tradeName
                bcRegistryId
              }
            }
          }
          ...SelectRfpWidget_query
          ...SelectProjectStatusWidget_query
        }
      }
    `,
    props.query
  );

  let selectedOperator = useMemo(() => {
    return query.allOperators.edges.find(
      ({ node }) => node.rowId === props.formData.operatorId
    );
  }, [query, props.formData.operatorId]);

  const schema: JSONSchema7 = useMemo(() => {
    return {
      type: "object",
      required: [
        "rfpNumber",
        "projectName",
        "summary",
        "operatorId",
        "fundingStreamRfpId",
      ],
      properties: {
        rfpNumber: {
          type: "string",
          title: "RFP Number",
          pattern: "^((\\d{4})-RFP-([1-2])-(\\d{3,4})-([A-Z]{4}))$",
        },
        projectName: { type: "string", title: "Project Name" },
        summary: { type: "string", title: "Summary" },
        operatorId: {
          type: "number",
          title: "Legal Operator Name and BC Registry ID",
          default: undefined,
          anyOf: query.allOperators.edges.map(({ node }) => {
            return {
              type: "number",
              title: `${node.legalName} (${node.bcRegistryId})`,
              enum: [node.rowId],
              value: node.rowId,
            };
          }),
        },
        operatorTradeName: {
          type: "string",
        },
        fundingStreamRfpId: {
          type: "number",
          title: "Funding Stream RFP ID",
          default: undefined,
        },
        projectStatusId: {
          type: "number",
          title: "Project Status",
          default: undefined,
        },
      },
    };
  }, [query]);

  const uiSchema = useMemo(() => {
    return {
      rfpNumber: {
        "ui:placeholder": "2020-RFP-1-456-ABCD",
        "ui:col-md": 12,
        "bcgov:size": "small",
      },
      projectName: {
        "ui:placeholder": "Short project name",
        "ui:col-md": 12,
        "bcgov:size": "small",
      },
      summary: {
        "ui:col-md": 12,
        "ui:widget": "TextAreaWidget",
        "bcgov:size": "small",
      },
      operatorId: {
        "ui:placeholder": "Select an Operator",
        "ui:col-md": 12,
        "bcgov:size": "small",
        "ui:widget": "SearchWidget",
      },
      operatorTradeName: {
        "ui:col-md": 12,
        "ui:widget": "DisplayOnly",
        "bcgov:size": "small",
        "ui:options": {
          text: `${selectedOperator ? selectedOperator.node.tradeName : ""}`,
          title: "Trade Name",
        },
      },
      fundingStreamRfpId: {
        "ui:widget": "SelectRfpWidget",
        "ui:col-md": 12,
        "bcgov:size": "small",
      },
      projectStatusId: {
        "ui:placeholder": "Select a Project Status",
        "ui:widget": "SelectProjectStatusWidget",
        "ui:col-md": 12,
        "bcgov:size": "small",
      },
    };
  }, [selectedOperator]);

  return (
    <FormBase
      {...props}
      schema={schema}
      uiSchema={uiSchema}
      formContext={{ query, form: props.formData }}
      widgets={{
        SelectRfpWidget: SelectRfpWidget,
        SelectProjectStatusWidget: SelectProjectStatusWidget,
      }}
    />
  );
};

export default ProjectForm;
