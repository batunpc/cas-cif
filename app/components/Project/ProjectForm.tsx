import type { JSONSchema7 } from "json-schema";
import FormBase from "../Form/FormBase";
import FormComponentProps from "../Form/FormComponentProps";
import { graphql, useFragment } from "react-relay";
import type { ProjectForm_query$key } from "__generated__/ProjectForm_query.graphql";
import { useMemo } from "react";

interface Props extends FormComponentProps {
  query: ProjectForm_query$key;
}

const ProjectForm: React.FC<Props> = (props) => {
  const { query } = useFragment(
    graphql`
      fragment ProjectForm_query on Query {
        query {
          allOperators {
            edges {
              node {
                rowId
                legalName
                tradeName
                bcRegistryId
              }
            }
          }
          allFundingStreams {
            edges {
              node {
                rowId
                name
                description
              }
            }
          }
        }
      }
    `,
    props.query
  );

  let selectedOperator = useMemo(() => {
      return query.allOperators.edges.find(
        ({ node }) => node.rowId === props.formData.operatorId
      )
    }, [query, props.formData.operatorId]);

  let selectedFundingStream = useMemo(() => {
      return query.allFundingStreams.edges.find(
        ({ node }) => node.rowId === props.formData.fundingStreamId
      )
    }, [query, props.formData.fundingStreamId]);

  const schema: JSONSchema7 = useMemo(() => {
    return {
      type: "object",
      required: ["rfpNumber", "description", "operatorId", "fundingStreamId"],
      properties: {
        rfpNumber: {
          type: "string",
          title: "RFP Number",
          pattern: "^((\\d{4})-RFP-([1-2])-(\\d{3,4})-([A-Z]{4}))$",
        },
        description: { type: "string", title: "Description" },
        operatorId: {
          type: "number",
          title: "Legal Operator Name and BC Registry ID",
          default: undefined,
          anyOf: query.allOperators.edges.map(({ node }) => {
            return {
              type: "number",
              title: `${node.legalName} (${node.bcRegistryId})`,
              enum: [node.rowId],
              value: node.rowId,
            };
          }),
        },
        operatorTradeName: {
          type: "string",
        },
        fundingStreamId: {
          type: "number",
          title: "Funding Stream ID",
          default: undefined,
          anyOf: query.allFundingStreams.edges.map(({ node }) => {
            return {
              type: "number",
              title: node.name,
              enum: [node.rowId],
              value: node.rowId,
            };
          }),
        },
        fundingStreamDescription: {
          type: "string",
        },
      },
    };
  }, [query]);

  const uiSchema = useMemo(() => {
    return {
      rfp_number: {
        "ui:placeholder": "2020-RFP-1-456-ABCD",
        "ui:col-md": 12,
        "bcgov:size": "small",
      },
      description: {
        "ui:placeholder": "describe the project...",
        "ui:col-md": 12,
        "bcgov:size": "small",
      },
      operatorId: {
        "ui:placeholder": "Select an Operator",
        "ui:col-md": 12,
        "bcgov:size": "small",
        "ui:widget": "SearchWidget",
      },
      operatorTradeName: {
        "ui:col-md": 12,
        "ui:widget": "DisplayOnly",
        "bcgov:size": "small",
        "ui:options": {
          text: `${selectedOperator ? selectedOperator.node.tradeName : ""}`,
          title: "Trade Name",
        },
      },
      fundingStreamId: {
        "ui:placeholder": "Select a Funding Stream",
        "ui:widget": "select",
        "ui:col-md": 12,
        "bcgov:size": "small",
      },
      fundingStreamDescription: {
        "ui:col-md": 12,
        "ui:widget": "DisplayOnly",
        "bcgov:size": "large",
        "ui:options": {
          text: `${
            selectedFundingStream ? selectedFundingStream.node.description : ""
          }`,
          title: "Description",
        },
      },
    };
  }, [selectedOperator, selectedFundingStream]);

  return <FormBase {...props} schema={schema} uiSchema={uiSchema} />;
};

export default ProjectForm;
