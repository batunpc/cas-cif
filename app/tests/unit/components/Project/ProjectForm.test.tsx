import { render, screen, fireEvent } from "@testing-library/react";
import FormComponentProps from "components/Form/FormComponentProps";
import ProjectForm from "components/Project/ProjectForm";
import { createMockEnvironment, MockPayloadGenerator } from "relay-test-utils";
import {
  RelayEnvironmentProvider,
  useLazyLoadQuery,
  graphql,
} from "react-relay";
import compiledProjectFormQuery from "__generated__/ProjectFormQuery.graphql";

const loadedQuery = graphql`
  query ProjectFormQuery @relay_test_operation {
    query {
      # Spread the fragment you want to test here
      ...ProjectForm_query
    }
  }
`;

const props: FormComponentProps = {
  formData: {},
  onChange: jest.fn(),
  onFormErrors: jest.fn(),
};

let environment;
const TestRenderer = () => {
  const data = useLazyLoadQuery(loadedQuery, {});
  return <ProjectForm {...props} query={data.query} />;
};
const renderProjectForm = () => {
  return render(
    <RelayEnvironmentProvider environment={environment}>
      <TestRenderer />
    </RelayEnvironmentProvider>
  );
}

describe("The Project Form", () => {
  beforeEach(() => {
    environment = createMockEnvironment();
  });
  it("matches the snapshot", () => {
    environment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation)
    );

    environment.mock.queuePendingOperation(compiledProjectFormQuery, {});

    const componentUnderTest = renderProjectForm();
    expect(componentUnderTest.container).toMatchSnapshot();
  });
  it("triggers the applyFormChange with the proper data", () => {
    const changeSpy = jest.fn();

    props.onChange = changeSpy;
    environment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation)
    );

    environment.mock.queuePendingOperation(compiledProjectFormQuery, {});

    renderProjectForm();
    fireEvent.change(screen.getByLabelText("RFP Number*"), {
      target: { value: "testidentifier" },
    });

    expect(changeSpy).toHaveBeenCalledWith({
      rfpNumber: "testidentifier",
      description: undefined,
    });
    changeSpy.mockClear();

    fireEvent.change(screen.getByLabelText("Description*"), {
      target: { value: "testdescription" },
    });

    expect(changeSpy).toHaveBeenCalledWith({
      rfpNumber: "testidentifier",
      description: "testdescription",
    });
  });
  it("loads with the correct initial form data", () => {
    props.formData = {
      rfpNumber: "12345678",
      description: "d",
      operatorId: 1,
    };
    environment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation, {
        Query() {
          return {
            allOperators: {
              edges: [
                {
                  node: {
                    rowId: 1,
                    legalName: "test operator",
                    bcRegistryId: "1234abcd",
                  },
                },
              ],
            },
          };
        },
      })
    );

    environment.mock.queuePendingOperation(compiledProjectFormQuery, {});

    renderProjectForm();

    expect(screen.getByLabelText("RFP Number*").value).toBe("12345678");
    expect(screen.getByLabelText("Description*").value).toBe("d");
    expect(screen.getByPlaceholderText("Select an Operator").value).toBe(
      "test operator (1234abcd)"
    );
  });
  it("calls onformerrors on first render if there are errors", () => {
    const onFormErrorsSpy = jest.fn();

    props.formData = {
      rfpNumber: "",
      description: "",
      operatorId: 1,
    };
    props.onFormErrors = onFormErrorsSpy;

    environment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation, {
        Query() {
          return {
            allOperators: {
              edges: [
                {
                  node: {
                    rowId: 1,
                    legalName: "test operator",
                    bcRegistryId: "1234abcd",
                  },
                },
              ],
            },
          };
        },
      })
    );

    environment.mock.queuePendingOperation(compiledProjectFormQuery, {});

    renderProjectForm();

    expect(onFormErrorsSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        rfpNumber: expect.anything(),
        description: null,
      })
    );
  });

  it("calls onformerrors with null if there are no errors", () => {
    const onFormErrorsSpy = jest.fn();

    props.formData = {
      rfpNumber: "1999-RFP-1-123-ABCD",
      description: "d",
      operatorId: 1,
    };
    props.onFormErrors = onFormErrorsSpy;
    environment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation, {
        Query() {
          return {
            allOperators: {
              edges: [
                {
                  node: {
                    rowId: 1,
                    legalName: "test operator",
                    bcRegistryId: "1234abcd",
                  },
                },
              ],
            },
          };
        },
      })
    );

    environment.mock.queuePendingOperation(compiledProjectFormQuery, {});

    renderProjectForm();

    expect(onFormErrorsSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        rfpNumber: null,
        description: null,
        operatorId: null,
      })
    );
  });

  it("calls onformerrors if a fields becomes empty", () => {
    const onFormErrorsSpy = jest.fn();

    props.formData = {
      rfpNumber: "1999-RFP-1-123-ABCD",
      description: "desc",
    };
    props.onFormErrors = onFormErrorsSpy;

    environment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation, {
        Query() {
          return {
            allOperators: {
              edges: [
                {
                  node: {
                    rowId: 1,
                    legalName: "test operator",
                    bcRegistryId: "1234abcd",
                  },
                },
              ],
            },
          };
        },
      })
    );

    environment.mock.queuePendingOperation(compiledProjectFormQuery, {});

    renderProjectForm();

    fireEvent.change(screen.getByLabelText("Description*"), {
      target: { value: "" },
    });

    expect(onFormErrorsSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        rfpNumber: null,
        description: { __errors: ["is a required property"] },
      })
    );
  });

  it("calls onformerrors if the project unique id doesnt match format", () => {
    const onFormErrorsSpy = jest.fn();

    props.formData = {
      rfpNumber: "1999123-RFP-1-123-ABCD",
      description: "desc",
      operatorId: 1,
    };
    props.onFormErrors = onFormErrorsSpy;

    environment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation, {
        Query() {
          return {
            allOperators: {
              edges: [
                {
                  node: {
                    rowId: 1,
                    legalName: "test operator",
                    bcRegistryId: "1234abcd",
                  },
                },
              ],
            },
          };
        },
      })
    );

    environment.mock.queuePendingOperation(compiledProjectFormQuery, {});

    renderProjectForm();

    expect(onFormErrorsSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        rfpNumber: expect.anything(),
        description: null,
        operatorId: null,
      })
    );
  });
});
