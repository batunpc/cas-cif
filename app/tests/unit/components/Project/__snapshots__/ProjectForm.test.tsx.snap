// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Project Form matches the snapshot 1`] = `
<div>
  <fieldset
    class="jsx-3408663875"
  >
    <legend
      class="jsx-3408663875"
    >
      Background
    </legend>
    <form
      class="rjsf"
    >
      <div
        class="panel panel-danger errors"
      >
        <div
          class="panel-heading"
        >
          <h3
            class="panel-title"
          >
            Errors
          </h3>
        </div>
        <ul
          class="list-group"
        >
          <li
            class="list-group-item text-danger"
          >
            .rfpNumber is a required property
          </li>
          <li
            class="list-group-item text-danger"
          >
            .description is a required property
          </li>
          <li
            class="list-group-item text-danger"
          >
            .operatorId is a required property
          </li>
        </ul>
      </div>
      <fieldset
        class="sc-dkPtRN bfJYwN pg-fieldset"
        id="root"
      >
        <div
          class="sc-hBUSln jEYZxL pg-input"
        >
          <label
            class="sc-fotOHu bhxnix pg-input-label"
            for="root_rfpNumber"
          >
            RFP Number*
          </label>
          <input
            aria-label="RFP Number*"
            class="sc-fFeiMQ jSxRsG pg-input-input"
            id="root_rfpNumber"
            name="root_rfpNumber-input"
            placeholder=""
            required=""
            value=""
          />
        </div>
        <div>
          <ul
            class="error-detail bs-callout bs-callout-info"
          >
            <li
              class="text-danger"
            >
              is a required property
            </li>
          </ul>
        </div>
        <div
          class="sc-hBUSln cGjSDY pg-input"
        >
          <label
            class="sc-fotOHu bhxnix pg-input-label"
            for="root_description"
          >
            Description*
          </label>
          <input
            aria-label="Description*"
            class="sc-fFeiMQ jSxRsG pg-input-input"
            id="root_description"
            name="root_description-input"
            placeholder="describe the project..."
            required=""
            value=""
          />
        </div>
        <div>
          <ul
            class="error-detail bs-callout bs-callout-info"
          >
            <li
              class="text-danger"
            >
              is a required property
            </li>
          </ul>
        </div>
        <label>
          Legal Operator Name and BC Registry ID*
        </label>
        <div
          aria-expanded="false"
          class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-symcig-MuiAutocomplete-root"
          role="combobox"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-autocomplete="list"
                aria-invalid="false"
                autocapitalize="none"
                autocomplete="off"
                class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                id="search-dropdown"
                placeholder="Select an Operator"
                spellcheck="false"
                type="text"
                value=""
              />
              <div
                class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
              >
                <button
                  aria-label="Clear"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
                  tabindex="-1"
                  title="Clear"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                    data-testid="CloseIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-label="Open"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                  tabindex="-1"
                  title="Open"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="ArrowDropDownIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-nnbavb"
                >
                  <span
                    class="notranslate"
                  >
                    â€‹
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div>
          <ul
            class="error-detail bs-callout bs-callout-info"
          >
            <li
              class="text-danger"
            >
              is a required property
            </li>
          </ul>
        </div>
        <span
          class="jsx-1332704472 paragraph-text"
        >
          <label
            class="jsx-1332704472"
          >
            Trade Name
          </label>
          <p
            class="jsx-1332704472"
          />
        </span>
      </fieldset>
    </form>
  </fieldset>
</div>
`;
